File: 72_llm_market_simulation/python/market/order_book.py
"""
Order Book Implementation for Market Simulation

A realistic limit order book with price-time priority matching.
Supports market orders, limit orders, and partial fills.
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import List, Optional, Tuple
import heapq
from datetime import datetime


class OrderType(Enum):
    """Type of order"""
    MARKET = "market"
    LIMIT = "limit"


class Side(Enum):
    """Side of the order (buy or sell)"""
    BUY = "buy"
    SELL = "sell"


@dataclass
class Order:
    """
    Represents a trading order

    Attributes:
        order_type: Type of order (market or limit)
        side: Side of the order (buy or sell)
        quantity: Number of shares to trade
        price: Limit price (None for market orders)
        agent_id: Identifier of the agent placing the order
        timestamp: Time when order was placed
        order_id: Unique identifier for the order
    """
    order_type: OrderType
    side: Side
    quantity: int
    price: Optional[float] = None
    agent_id: str = ""
    timestamp: int = 0
    order_id: int = 0


@dataclass
class OrderResult:
    """
    Result of submitting an order

    Attributes:
        filled_quantity: Number of shares filled
        average_price: Average price of fills
        remaining_quantity: Number of shares not filled
        status: Status of the order ("filled", "partial", "pending", "rejected")
        fills: List of individual fills (price, quantity)
    """
    filled_quantity: int
    average_price: float
    remaining_quantity: int
    status: str
    fills: List[Tuple[float, int]] = field(default_factory=list)


@dataclass
class BookLevel:
    """Represents a price level in the order book"""
    price: float
    orders: List[Tuple[int, int, str]]  # (order_id, quantity, agent_id)


class OrderBook:
    """
    Limit Order Book with Price-Time Priority

    Implements a realistic order book that matches orders based on
    price-time priority. Supports limit orders, market orders,
    partial fills, and order cancellation.

    Examples:
        >>> book = OrderBook(tick_size=0.01)
        >>> order = Order(OrderType.LIMIT, Side.BUY, 100, price=50.0, agent_id="agent1")
        >>> result = book.add_limit_order(order)
        >>> print(result.status)
        'pending'
    """

    def __init__(self, tick_size: float = 0.01):
        """
        Initialize order book

        Args:
            tick_size: Minimum price increment
        """
        self.tick_size = tick_size
        self.bids: List[Tuple[float, int, int, str]] = []  # (-price, timestamp, qty, agent_id)
        self.asks: List[Tuple[float, int, int, str]] = []  # (price, timestamp, qty, agent_id)
        self.order_id_counter = 0
        self.timestamp_counter = 0
        self.trade_history: List[dict] = []

    def _next_order_id(self) -> int:
        """Get next unique order ID"""
        self.order_id_counter += 1
        return self.order_id_counter

    def _next_timestamp(self) -> int:
        """Get next timestamp for ordering"""
        self.timestamp_counter += 1
        return self.timestamp_counter

    def add_limit_order(self, order: Order) -> OrderResult:
        """
        Add a limit order to the book

        Args:
            order: Limit order to add

        Returns:
            OrderResult with fill information
        """
        if order.order_type != OrderType.LIMIT or order.price is None:
            return OrderResult(0, 0.0, order.quantity, "rejected", [])

        order.order_id = self._next_order_id()
        order.timestamp = self._next_timestamp()

        if order.side == Side.BUY:
            return self._process_buy_limit(order)
        else:
            return self._process_sell_limit(order)

    def _process_buy_limit(self, order: Order) -> OrderResult:
        """Process a buy limit order"""
        fills = []
        filled_qty = 0
        remaining = order.quantity

        # Try to match against asks
        while remaining > 0 and self.asks:
            best_ask = self.asks[0]
            if order.price >= best_ask[0]:  # Can trade
                trade_qty = min(remaining, best_ask[2])
                trade_price = best_ask[0]

                fills.append((trade_price, trade_qty))
                filled_qty += trade_qty
                remaining -= trade_qty

                # Record trade
                self.trade_history.append({
                    "price": trade_price,
                    "quantity": trade_qty,
                    "buyer": order.agent_id,
                    "seller": best_ask[3],
                    "timestamp": order.timestamp
                })

                # Update or remove the ask
                if trade_qty >= best_ask[2]:
                    heapq.heappop(self.asks)
                else:
                    # Update remaining quantity
                    self.asks[0] = (best_ask[0], best_ask[1], best_ask[2] - trade_qty, best_ask[3])
            else:
                break

        # Add remaining to bid book
        if remaining > 0:
            heapq.heappush(self.bids, (-order.price, order.timestamp, remaining, order.agent_id))

        # Calculate average price
        avg_price = sum(p * q for p, q in fills) / filled_qty if filled_qty > 0 else 0.0

        if remaining == 0:
            status = "filled"
        elif filled_qty > 0:
            status = "partial"
        else:
            status = "pending"

        return OrderResult(filled_qty, avg_price, remaining, status, fills)

    def _process_sell_limit(self, order: Order) -> OrderResult:
        """Process a sell limit order"""
        fills = []
        filled_qty = 0
        remaining = order.quantity

        # Try to match against bids
        while remaining > 0 and self.bids:
            best_bid = self.bids[0]
            bid_price = -best_bid[0]  # Convert back from negative
            if order.price <= bid_price:  # Can trade
                trade_qty = min(remaining, best_bid[2])
                trade_price = bid_price

                fills.append((trade_price, trade_qty))
                filled_qty += trade_qty
                remaining -= trade_qty

                # Record trade
                self.trade_history.append({
                    "price": trade_price,
                    "quantity": trade_qty,
                    "buyer": best_bid[3],
                    "seller": order.agent_id,
                    "timestamp": order.timestamp
                })

                # Update or remove the bid
                if trade_qty >= best_bid[2]:
                    heapq.heappop(self.bids)
                else:
                    # Update remaining quantity
                    self.bids[0] = (best_bid[0], best_bid[1], best_bid[2] - trade_qty, best_bid[3])
            else:
                break

        # Add remaining to ask book
        if remaining > 0:
            heapq.heappush(self.asks, (order.price, order.timestamp, remaining, order.agent_id))

        # Calculate average price
        avg_price = sum(p * q for p, q in fills) / filled_qty if filled_qty > 0 else 0.0

        if remaining == 0:
            status = "filled"
        elif filled_qty > 0:
            status = "partial"
        else:
            status = "pending"

        return OrderResult(filled_qty, avg_price, remaining, status, fills)

    def execute_market_order(self, order: Order) -> OrderResult:
        """
        Execute a market order immediately

        Args:
            order: Market order to execute

        Returns:
            OrderResult with fill information
        """
        order.order_id = self._next_order_id()
        order.timestamp = self._next_timestamp()

        if order.side == Side.BUY:
            return self._execute_market_buy(order)
        else:
            return self._execute_market_sell(order)

    def _execute_market_buy(self, order: Order) -> OrderResult:
        """Execute a market buy order"""
        fills = []
        filled_qty = 0
        remaining = order.quantity

        while remaining > 0 and self.asks:
            best_ask = self.asks[0]
            trade_qty = min(remaining, best_ask[2])
            trade_price = best_ask[0]

            fills.append((trade_price, trade_qty))
            filled_qty += trade_qty
            remaining -= trade_qty

            self.trade_history.append({
                "price": trade_price,
                "quantity": trade_qty,
                "buyer": order.agent_id,
                "seller": best_ask[3],
                "timestamp": order.timestamp
            })

            if trade_qty >= best_ask[2]:
                heapq.heappop(self.asks)
            else:
                self.asks[0] = (best_ask[0], best_ask[1], best_ask[2] - trade_qty, best_ask[3])

        avg_price = sum(p * q for p, q in fills) / filled_qty if filled_qty > 0 else 0.0
        status = "filled" if remaining == 0 else ("partial" if filled_qty > 0 else "rejected")

        return OrderResult(filled_qty, avg_price, remaining, status, fills)

    def _execute_market_sell(self, order: Order) -> OrderResult:
        """Execute a market sell order"""
        fills = []
        filled_qty = 0
        remaining = order.quantity

        while remaining > 0 and self.bids:
            best_bid = self.bids[0]
            bid_price = -best_bid[0]
            trade_qty = min(remaining, best_bid[2])

            fills.append((bid_price, trade_qty))
            filled_qty += trade_qty
            remaining -= trade_qty

            self.trade_history.append({
                "price": bid_price,
                "quantity": trade_qty,
                "buyer": best_bid[3],
                "seller": order.agent_id,
                "timestamp": order.timestamp
            })

            if trade_qty >= best_bid[2]:
                heapq.heappop(self.bids)
            else:
                self.bids[0] = (best_bid[0], best_bid[1], best_bid[2] - trade_qty, best_bid[3])

        avg_price = sum(p * q for p, q in fills) / filled_qty if filled_qty > 0 else 0.0
        status = "filled" if remaining == 0 else ("partial" if filled_qty > 0 else "rejected")

        return OrderResult(filled_qty, avg_price, remaining, status, fills)

    def get_best_bid(self) -> Optional[float]:
        """Get best bid price"""
        if self.bids:
            return -self.bids[0][0]
        return None

    def get_best_ask(self) -> Optional[float]:
        """Get best ask price"""
        if self.asks:
            return self.asks[0][0]
        return None

    def get_spread(self) -> Optional[float]:
        """Get bid-ask spread"""
        bid, ask = self.get_best_bid(), self.get_best_ask()
        if bid is not None and ask is not None:
            return ask - bid
        return None

    def get_mid_price(self) -> Optional[float]:
        """Get mid-market price"""
        bid, ask = self.get_best_bid(), self.get_best_ask()
        if bid is not None and ask is not None:
            return (bid + ask) / 2
        return None

    def get_bid_depth(self, levels: int = 5) -> List[Tuple[float, int]]:
        """Get bid side depth (price, total quantity)"""
        depth = {}
        for neg_price, _, qty, _ in self.bids:
            price = -neg_price
            depth[price] = depth.get(price, 0) + qty

        sorted_depth = sorted(depth.items(), reverse=True)[:levels]
        return sorted_depth

    def get_ask_depth(self, levels: int = 5) -> List[Tuple[float, int]]:
        """Get ask side depth (price, total quantity)"""
        depth = {}
        for price, _, qty, _ in self.asks:
            depth[price] = depth.get(price, 0) + qty

        sorted_depth = sorted(depth.items())[:levels]
        return sorted_depth

    def get_last_trade_price(self) -> Optional[float]:
        """Get the price of the last trade"""
        if self.trade_history:
            return self.trade_history[-1]["price"]
        return None

    def clear(self):
        """Clear all orders from the book"""
        self.bids.clear()
        self.asks.clear()


if __name__ == "__main__":
    # Example usage
    book = OrderBook(tick_size=0.01)

    # Add some limit orders
    buy_order = Order(OrderType.LIMIT, Side.BUY, 100, price=99.50, agent_id="buyer1")
    result = book.add_limit_order(buy_order)
    print(f"Buy order: {result.status}")

    sell_order = Order(OrderType.LIMIT, Side.SELL, 50, price=100.50, agent_id="seller1")
    result = book.add_limit_order(sell_order)
    print(f"Sell order: {result.status}")

    print(f"Best bid: {book.get_best_bid()}")
    print(f"Best ask: {book.get_best_ask()}")
    print(f"Spread: {book.get_spread()}")

    # Execute a market order
    market_buy = Order(OrderType.MARKET, Side.BUY, 30, agent_id="market_buyer")
    result = book.execute_market_order(market_buy)
    print(f"Market buy: filled {result.filled_quantity} @ {result.average_price:.2f}")

File: 72_llm_market_simulation/rust_market_sim/src/market/order_book.rs
//! Order Book Implementation
//!
//! A limit order book with price-time priority matching algorithm.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::BinaryHeap;
use std::cmp::Ordering;
use uuid::Uuid;

/// Order side (buy or sell)
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Side {
    Buy,
    Sell,
}

/// Order type
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum OrderType {
    /// Market order - executes immediately at best available price
    Market,
    /// Limit order - executes only at specified price or better
    Limit,
}

/// A single order in the order book
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Order {
    /// Unique order ID
    pub id: String,
    /// Agent who placed the order
    pub agent_id: String,
    /// Order side
    pub side: Side,
    /// Order type
    pub order_type: OrderType,
    /// Limit price (for limit orders)
    pub price: f64,
    /// Order quantity
    pub quantity: u64,
    /// Remaining quantity
    pub remaining: u64,
    /// Order timestamp
    pub timestamp: DateTime<Utc>,
}

impl Order {
    /// Create a new order
    pub fn new(
        agent_id: String,
        side: Side,
        order_type: OrderType,
        price: f64,
        quantity: u64,
    ) -> Self {
        Self {
            id: Uuid::new_v4().to_string(),
            agent_id,
            side,
            order_type,
            price,
            quantity,
            remaining: quantity,
            timestamp: Utc::now(),
        }
    }

    /// Check if order is fully filled
    pub fn is_filled(&self) -> bool {
        self.remaining == 0
    }
}

// For buy orders: higher price = higher priority
// For sell orders: lower price = higher priority
// At same price: earlier timestamp = higher priority

#[derive(Debug, Clone)]
struct BidEntry(Order);

#[derive(Debug, Clone)]
struct AskEntry(Order);

impl PartialEq for BidEntry {
    fn eq(&self, other: &Self) -> bool {
        self.0.id == other.0.id
    }
}

impl Eq for BidEntry {}

impl PartialOrd for BidEntry {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for BidEntry {
    fn cmp(&self, other: &Self) -> Ordering {
        // Higher price first, then earlier time
        match self.0.price.partial_cmp(&other.0.price) {
            Some(Ordering::Equal) => other.0.timestamp.cmp(&self.0.timestamp),
            Some(ord) => ord,
            None => Ordering::Equal,
        }
    }
}

impl PartialEq for AskEntry {
    fn eq(&self, other: &Self) -> bool {
        self.0.id == other.0.id
    }
}

impl Eq for AskEntry {}

impl PartialOrd for AskEntry {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for AskEntry {
    fn cmp(&self, other: &Self) -> Ordering {
        // Lower price first, then earlier time
        match other.0.price.partial_cmp(&self.0.price) {
            Some(Ordering::Equal) => other.0.timestamp.cmp(&self.0.timestamp),
            Some(ord) => ord,
            None => Ordering::Equal,
        }
    }
}

/// Result of order execution
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrderResult {
    /// Order ID
    pub order_id: String,
    /// Whether order was accepted
    pub accepted: bool,
    /// Filled quantity
    pub filled_quantity: u64,
    /// Average fill price
    pub average_price: f64,
    /// List of fills
    pub fills: Vec<Fill>,
    /// Remaining quantity in book
    pub remaining: u64,
}

/// A single fill (trade execution)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Fill {
    /// Price at which trade occurred
    pub price: f64,
    /// Quantity traded
    pub quantity: u64,
    /// Counterparty agent ID
    pub counterparty: String,
    /// Timestamp
    pub timestamp: DateTime<Utc>,
}

/// Limit Order Book
#[derive(Debug)]
pub struct OrderBook {
    /// Buy orders (bids)
    bids: BinaryHeap<BidEntry>,
    /// Sell orders (asks)
    asks: BinaryHeap<AskEntry>,
    /// Last traded price
    pub last_price: f64,
    /// Trade history
    trades: Vec<Fill>,
}

impl OrderBook {
    /// Create a new empty order book
    pub fn new(initial_price: f64) -> Self {
        Self {
            bids: BinaryHeap::new(),
            asks: BinaryHeap::new(),
            last_price: initial_price,
            trades: Vec::new(),
        }
    }

    /// Get best bid price
    pub fn best_bid(&self) -> Option<f64> {
        self.bids.peek().map(|e| e.0.price)
    }

    /// Get best ask price
    pub fn best_ask(&self) -> Option<f64> {
        self.asks.peek().map(|e| e.0.price)
    }

    /// Get bid-ask spread
    pub fn spread(&self) -> Option<f64> {
        match (self.best_ask(), self.best_bid()) {
            (Some(ask), Some(bid)) => Some(ask - bid),
            _ => None,
        }
    }

    /// Get mid price
    pub fn mid_price(&self) -> Option<f64> {
        match (self.best_ask(), self.best_bid()) {
            (Some(ask), Some(bid)) => Some((ask + bid) / 2.0),
            _ => None,
        }
    }

    /// Submit an order to the book
    pub fn submit_order(&mut self, mut order: Order) -> OrderResult {
        let mut result = OrderResult {
            order_id: order.id.clone(),
            accepted: true,
            filled_quantity: 0,
            average_price: 0.0,
            fills: Vec::new(),
            remaining: order.remaining,
        };

        match order.side {
            Side::Buy => self.match_buy_order(&mut order, &mut result),
            Side::Sell => self.match_sell_order(&mut order, &mut result),
        }

        // Capture remaining before potentially moving order
        let remaining = order.remaining;

        // Add remaining quantity to book (for limit orders)
        if remaining > 0 && order.order_type == OrderType::Limit {
            match order.side {
                Side::Buy => self.bids.push(BidEntry(order)),
                Side::Sell => self.asks.push(AskEntry(order)),
            }
        }

        // Calculate average price
        if result.filled_quantity > 0 {
            let total_value: f64 = result.fills.iter()
                .map(|f| f.price * f.quantity as f64)
                .sum();
            result.average_price = total_value / result.filled_quantity as f64;
            self.last_price = result.fills.last().map(|f| f.price).unwrap_or(self.last_price);
        }

        result.remaining = remaining;
        result
    }

    fn match_buy_order(&mut self, order: &mut Order, result: &mut OrderResult) {
        let mut matched_asks = Vec::new();

        while order.remaining > 0 {
            if let Some(mut ask_entry) = self.asks.pop() {
                let can_match = match order.order_type {
                    OrderType::Market => true,
                    OrderType::Limit => order.price >= ask_entry.0.price,
                };

                if can_match {
                    let fill_qty = order.remaining.min(ask_entry.0.remaining);
                    let fill_price = ask_entry.0.price;

                    let fill = Fill {
                        price: fill_price,
                        quantity: fill_qty,
                        counterparty: ask_entry.0.agent_id.clone(),
                        timestamp: Utc::now(),
                    };

                    order.remaining -= fill_qty;
                    ask_entry.0.remaining -= fill_qty;
                    result.filled_quantity += fill_qty;
                    result.fills.push(fill.clone());
                    self.trades.push(fill);

                    if ask_entry.0.remaining > 0 {
                        matched_asks.push(ask_entry);
                    }
                } else {
                    matched_asks.push(ask_entry);
                    break;
                }
            } else {
                break;
            }
        }

        // Put back unmatched asks
        for ask in matched_asks {
            self.asks.push(ask);
        }
    }

    fn match_sell_order(&mut self, order: &mut Order, result: &mut OrderResult) {
        let mut matched_bids = Vec::new();

        while order.remaining > 0 {
            if let Some(mut bid_entry) = self.bids.pop() {
                let can_match = match order.order_type {
                    OrderType::Market => true,
                    OrderType::Limit => order.price <= bid_entry.0.price,
                };

                if can_match {
                    let fill_qty = order.remaining.min(bid_entry.0.remaining);
                    let fill_price = bid_entry.0.price;

                    let fill = Fill {
                        price: fill_price,
                        quantity: fill_qty,
                        counterparty: bid_entry.0.agent_id.clone(),
                        timestamp: Utc::now(),
                    };

                    order.remaining -= fill_qty;
                    bid_entry.0.remaining -= fill_qty;
                    result.filled_quantity += fill_qty;
                    result.fills.push(fill.clone());
                    self.trades.push(fill);

                    if bid_entry.0.remaining > 0 {
                        matched_bids.push(bid_entry);
                    }
                } else {
                    matched_bids.push(bid_entry);
                    break;
                }
            } else {
                break;
            }
        }

        // Put back unmatched bids
        for bid in matched_bids {
            self.bids.push(bid);
        }
    }

    /// Get total bid volume at top N levels
    pub fn bid_volume(&self, levels: usize) -> u64 {
        self.bids.iter()
            .take(levels)
            .map(|e| e.0.remaining)
            .sum()
    }

    /// Get total ask volume at top N levels
    pub fn ask_volume(&self, levels: usize) -> u64 {
        self.asks.iter()
            .take(levels)
            .map(|e| e.0.remaining)
            .sum()
    }

    /// Get trade history
    pub fn get_trades(&self) -> &[Fill] {
        &self.trades
    }

    /// Clear all orders
    pub fn clear(&mut self) {
        self.bids.clear();
        self.asks.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_limit_order_matching() {
        let mut book = OrderBook::new(100.0);

        // Add sell limit order
        let sell_order = Order::new(
            "seller".to_string(),
            Side::Sell,
            OrderType::Limit,
            101.0,
            10,
        );
        book.submit_order(sell_order);

        assert_eq!(book.best_ask(), Some(101.0));
        assert_eq!(book.best_bid(), None);

        // Add buy limit order that matches
        let buy_order = Order::new(
            "buyer".to_string(),
            Side::Buy,
            OrderType::Limit,
            101.0,
            5,
        );
        let result = book.submit_order(buy_order);

        assert_eq!(result.filled_quantity, 5);
        assert_eq!(result.average_price, 101.0);
    }

    #[test]
    fn test_market_order() {
        let mut book = OrderBook::new(100.0);

        // Add liquidity
        book.submit_order(Order::new(
            "mm".to_string(),
            Side::Sell,
            OrderType::Limit,
            100.5,
            100,
        ));

        // Market buy
        let result = book.submit_order(Order::new(
            "buyer".to_string(),
            Side::Buy,
            OrderType::Market,
            0.0, // Price ignored for market orders
            50,
        ));

        assert_eq!(result.filled_quantity, 50);
        assert_eq!(book.last_price, 100.5);
    }

    #[test]
    fn test_spread() {
        let mut book = OrderBook::new(100.0);

        book.submit_order(Order::new(
            "mm".to_string(),
            Side::Buy,
            OrderType::Limit,
            99.5,
            100,
        ));

        book.submit_order(Order::new(
            "mm".to_string(),
            Side::Sell,
            OrderType::Limit,
            100.5,
            100,
        ));

        assert_eq!(book.spread(), Some(1.0));
        assert_eq!(book.mid_price(), Some(100.0));
    }
}
